### DISCUSSION TITLE
Consultas sobre ejercicio QMP

### DISCUSSION THREAD

**Discussion by Gonzalo-Avila** [dds-jv/foro#126] [Consultas de objetos] [answered]
Buenas! 

Tengo algunas dudas sobre lo que se espera de la cuarta iteración (tercera en realidad) de QMP:

1) ¿Cómo funciona la API de AccuWheather?
*  En primer lugar no me compila, me tira los siguientes errores:
	``` Type mismatch: cannot convert from List<new HashMap(){}> to List<Map<String,Object>>```
	```HashMap is a raw type. References to generic type HashMap<K,V> should be parameterized```
* Entiendo que es un mock, ya que tiene hardcodeado un return independientemente de lo que se envíe como parámetro. O sea, imita la salida de la API real cuando se le manda el mensaje getWeather, ¿es correcto? 
* Para usarla, entiendo que el get("Cadena") obtiene el valor de la lista asociado a la cadena, pero no entiendo que hace el get(0) anterior. Ejemplo: ```condicionesClimaticas.get(0).get("PrecipitationProbability");//Devuelve un número del 0 al 1```
* Relacionando con los conceptos que vimos en clase, esta API sería la interfaz que me brinda el componente externo para comunicarme con el, ¿verdad? En mi dominio me interesa poder enviar un mensaje obtenerTemperatura(Ciudad ciudad) (interfaz saliente) y recibir un entero , entonces necesito un adapter que se comunique con esa API en particular y me devuelva lo que quiero, ¿es así?
2) Respecto a los requerimientos: 

> Como usuarie de QuéMePongo, quiero poder recibir sugerencias de atuendos que tengan una prenda para cada categoría, aunque a futuro podrán tener más (Ej.: Una remera, un pantalón, zapatos y un gorro).

>Como usuarie de QuéMePongo, quiero que al generar una sugerencia las prendas sean acordes a la temperatura actual sabiendo que para cada prenda habrá una temperatura hasta la cual es adecuada. (Ej.: “Remera de mangas largas” no es apta a más de 20°C)

¿Se espera que codifiquemos algún algoritmo de aleatoriedad, o solo que estén las condiciones 
planteadas para que ese algoritmo se pueda realizar? (Como poder saber la temperatura máxima que admite un tipo de prenda y la temperatura actual para poder comparar)

Gracias y perdón por el choclo! 
Saludos.

**Comment by gastonprieto** [comment]
> En primer lugar no me compila, me tira los siguientes errores:
Type mismatch: cannot convert from List<new HashMap(){}> to List<Map<String,Object>>
HashMap is a raw type. References to generic type HashMap<K,V> should be parameterized

Ahí subi un cambio, en teoría igual debería haberte andando. Chequea que estes compilando con java 8.

> Entiendo que es un mock, ya que tiene hardcodeado un return independientemente de lo que se envíe como parámetro. O sea, imita la salida de la API real cuando se le manda el mensaje getWeather, ¿es correcto?

Esta fijo, para no hacer algo complejo, pero la idea es que asumas que ese objeto es como el que consulta a la API real. Por lo tanto no podes tomarlo como si fuera un impostor.

>Para usarla, entiendo que el get("Cadena") obtiene el valor de la lista asociado a la cadena, pero no entiendo que hace el get(0) anterior. Ejemplo: condicionesClimaticas.get(0).get("PrecipitationProbability");//Devuelve un número del 0 al 1

La estructura que te devuelven es una lista de maps `List<Map<String, Object>>` por eso el get(0) porque se trae el primer elemento de dicha lista que es un Map

Dictionary/Map, en [wollok](https://www.wollok.org/documentacion/conceptos/#diccionarios) en [Java](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html)

>Relacionando con los conceptos que vimos en clase, esta API sería la interfaz que me brinda el componente externo para comunicarme con el, ¿verdad? 

Si, es un componente externo que sirve para comunicarte con otro sistema.

> En mi dominio me interesa poder enviar un mensaje obtenerTemperatura(Ciudad ciudad) (interfaz saliente) y recibir un entero , entonces necesito un adapter que se comunique con esa API en particular y me devuelva lo que quiero, ¿es así?

Eso ya es parte de tu solución lo que quieras modelar :P

> ¿Se espera que codifiquemos algún algoritmo de aleatoriedad, o solo que estén las condiciones
planteadas para que ese algoritmo se pueda realizar?

Acá @roli93 va a poder poner mas luz sobre lo que se espera de la entrega.

**Comment by roli93** [comment]
Buenas! Recién mandé un par de mails que creo que deberían terminar de saldar tus dudas. Cualquier cosa, avisá por acá.

**Comment by Gonzalo-Avila** [comment]
Hola!

Gracias a ambos por las respuestas, me quedó mas claro. 

Aunque el último correo me genera otra duda: ¿no debería ser el generador de sugerencias quien se encargue de filtrar las prendas? Porque sino, no estaría entendiendo quien es el encargado de hacer esta tarea por el, y quien lo llama. 

O sea, en mi cabeza el uso seria:
**generador.generarSugerenciaDesde(obtenedorDeClima, ciudad, prendas)**

¿En que me estoy equivocando?

**Comment by roli93** [comment]
Ojo, GeneradorDeSugerencias es una **interfaz**, que la puede implementar cualquier objeto de tu sistema (el usuario, por ejemplo), y que sea el mismo que el que filtra. Probablemente tener un componente independiente para esta tarea sea una mala idea (sería un "objeto manejador", de esos que rompen el encapsulamiento)

**Comment by Gonzalo-Avila** [comment]
Aaah, genial. Entonces sería valido que sea el mismo objeto al que le pido la sugerencia, ¿no?

Se me ocurre esto:

![image](https://user-images.githubusercontent.com/44387608/82952504-7624af00-9f7f-11ea-8358-89f2ce89b8d2.png)

Y podría haber generadores por distintos criterios, todos ellos implementando la interfaz ¿Es razonable?

**Comment by roli93** [comment]
Podría ser. Aunque, nuevamente, ojo con poner ese comportamiento en un objeto separado. No existe ningún objeto de dominio donde tenga sentido y puedas poner eso? Qué objeto de tu dominio es el que tiene el acceso a las prendas y sabe en qué ciudad estás?

**Comment by Gonzalo-Avila** [comment]
Actualmente ninguno, pero se me ocurre que podría crear una clase Usuario que tenga una ciudad y un guardarropas. En tal caso generaría sugerencias para un usuario especifico en un momento especifico ¿puede ser?

**Comment by roli93** [comment]
Bien. Sí, claramente el usuario es el que tiene las prendas, (si no, donde están? :P). La abstracción de guardarropas tiene sentido también, aunque fijate cuán necesaria es para esta iteración. Y sí, también tendría sentido que el usuario conozca su ciudad. Por último, sí, las sugerencias siempre son para un determinado usuario en una determinada ciudad :P

**Comment by Gonzalo-Avila** [comment, marked as answer]
Perfecto, ahora si. Muchas gracias!

Cierro el issue.
